lib/auth/access.go

```diff
@@ -144,3 +148,71 @@ func (a *Server) DeleteLock(ctx context.Context, lockName string) error {
 	}
 	return nil
 }
+
+// checkRoleRulesConstraint checks if the request will result in having
+// no roles with rules to upsert roles.
+func (s *Server) checkRoleRulesConstraint(ctx context.Context, targetRole types.Role, request string) error {
+	rolesWithUpdateRolesRule, err := s.getRolesWithUpdateRolesRule(ctx)
+	if err != nil {
+		return trace.Wrap(err)
+	}
```

Comment:
Let's avoid fetching/filtering all of the roles in the cluster until after we check that this edit actually removes an update roles rule

Reply:
Updated